File: java/src/LeetCode/LC1129_ShortestPathWithAlternatingColors.java
package LeetCode;

import java.util.*;

public class LC1129_ShortestPathWithAlternatingColors {

    public int[] shortestAlternatingPaths(int n, int[][] redEdges, int[][] blueEdges) {
        // 创建图结构
        // graph[i] 表示 from 是 i， to 是添加进来的元素
        List<List<int[]>> graph = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            graph.add(new ArrayList<>());
        }
        for (int[] edge : redEdges) {
            int from = edge[0];
            int to = edge[1];
            // 创建从 from 指向 to 的图结构
            graph.get(from).add(new int[]{to, 0}); // 0 表示红色
        }
        for (int[] edge : blueEdges) {
            int from = edge[0];
            int to = edge[1];
            // 创建从 from 指向 to 的图结构
            graph.get(from).add(new int[]{to, 1}); // 1 表示蓝色
        } // 图结构创建完成

        boolean[][] visited = new boolean[n][2];
        Queue<int[]> queue = new LinkedList<>();

        queue.offer(new int[]{0, 0, 0}); // 三元组 {curNode, distance, color}
        queue.offer(new int[]{0, 0, 1});
        visited[0][0] = true;
        visited[0][1] = true;

        int[] ans = new int[n];
        Arrays.fill(ans, Integer.MAX_VALUE);
        ans[0] = 0;

        while (!queue.isEmpty()) {
            int[] arr = queue.poll();
            int curNode = arr[0];
            int distanceToStart = arr[1];
            int curNodeColor = arr[2];
            for (int[] it : graph.get(curNode)) {
                int nextNode = it[0];
                int nextNodeColor = it[1];
                if (curNodeColor != nextNodeColor && !visited[nextNode][nextNodeColor]) {
                    ans[nextNode] = Math.min(ans[nextNode], distanceToStart + 1);
                    queue.offer(new int[]{nextNode, distanceToStart + 1, nextNodeColor});
                    visited[nextNode][nextNodeColor] = true;
                }
            }
        }

        for (int i = 0; i < n; i++) {
            if (ans[i] == Integer.MAX_VALUE)
                ans[i] = -1;
        }

        return ans;
    }

}

File: cpp/Exercise/LC1129_ShortestPathWithAlternatingColors.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

class LC1129_ShortestPathWithAlternatingColors {
public:
    vector<int> shortestAlternatingPaths(int n, vector<vector<int>> &redEdges, vector<vector<int>> &blueEdges) {

        vector<vector<pair<int, int>>> graph(n); // 三元组
        vector<vector<int>> visited(n, vector<int>(2, 0)); // 二元组

        for (auto &edge: redEdges) {
            int from = edge[0];
            int to = edge[1];
            // 一个 from 可能对应多个 to 节点，每个 to 节点都有一个颜色属性
            graph[from].push_back({to, 0});
        }

        for (auto &edge: blueEdges) {
            int from = edge[0];
            int to = edge[1];
            graph[from].push_back({to, 1});
        }

        queue<vector<int>> q;
        vector<int> dis(n, INT_MAX);

        dis[0] = 0;
        q.push({0, 0, 1}); // 节点 距离 颜色
        q.push({0, 0, 0});

        while (!q.empty()) {
            // 当前节点
            auto it = q.front();
            int from = it[0];       // 当前节点的值
            int distance = it[1];   // 当前节点与 0 的距离
            int curColor = it[2];   // 当前节点的颜色

            q.pop();
            for (auto &[to, nextColor]: graph[from]) {
                if (nextColor == !curColor && !visited[to][nextColor]) {
                    dis[to] = min(dis[to], distance + 1);
                    q.push({to, distance + 1, nextColor});
                    visited[to][nextColor] = 1;
                }
            }
        }

        for (int i = 0; i < n; i++) {
            if (dis[i] == INT_MAX) {
                dis[i] = -1;
            }
        }

        return dis;
    }
};
